### ISSUE TITLE
Feature Request: Prevent OS screen-lock or sleep when using a gamepad

### ISSUE THREAD

**Issue by j-h-a** [subsoap/defos#145]
## Description

Most OSes have a screen-lock or sleep feature that activates when the user does not provide input for some time. When using a gamepad the OS does not recognise this input. It would be great to have a single cross-platform function in Defos to prevent this behaviour across macOS, Windows, and Linux.

## Proposed solution

Tell the OS not to allow sleeping while the game is active. Co-pilot reckons these OS functions are how to do it:

Implementation research:

* macOS: [`IOPMAssertionCreateWithName`](https://developer.apple.com/documentation/iokit/1557134-iopmassertioncreatewithname) while using [this assertion type](https://developer.apple.com/documentation/iokit/kiopmassertiontypepreventuseridledisplaysleep).  Suggest the assertion name come from window title also set by defos or default project name from Defold.
* Windows: [`SetThreadExecutionState`](https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-setthreadexecutionstate). Probably with the `ES_DISPLAY_REQUIRED` flag, but maybe the `ES_SYSTEM_REQUIRED` one - not sure.
* Linux: [`XScreenSaverSuspend`](https://www.x.org/releases/X11R7.6/doc/man/man3/XScreenSaverSuspend.3.xhtml) with `suspend` set to `false` to reset the timer.

The Windows and Linux ones appear to reset the idle timer, whereas the macOS one sets your requested assertion off/on. In order to present a consistent API to defos users, it might make sense to have them call a single function periodically and define a short time (20s or something) that if not called it will turn off the macOS one, and for Windows and Linux it can just pass on the call to reset the idle timer. This way it's the game developer's responsibility to keep calling it periodically while receiving gamepad input, then stop calling it to allow the system to sleep if the user really has gone away.

**Comment by j-h-a** [comment]
Hrm... I wonder if this is something Defold should do directly, and automatically, rather than in an extension 🤔? It's not something that really should need any developer-interaction, and the engine knows when there are inputs, etc.

**Comment by britzl** [comment]
> Hrm... I wonder if this is something Defold should do directly, and automatically, rather than in an extension 🤔? It's not something that really should need any developer-interaction, and the engine knows when there are inputs, etc.

Yes, this feels like something which should be available as a game.project option in desktop builds of Defold.

**Comment by subsoap** [comment]
WIP branch for this feature needs testing on each platform I only tested on macos/html5 for crashing 

https://github.com/subsoap/defos/tree/copilot/fix-screen-dimming-issue
